src/transform/src/nonnull_requirements.rs

```diff
@@ -143,11 +143,35 @@ impl NonNullRequirements {
                 self.action(input, columns, gets);
             }
             MirRelationExpr::Filter { input, predicates } => {
-                for predicate in predicates {
+                for predicate in predicates.iter() {
                     predicate.non_null_requirements(&mut columns);
                     // TODO: Not(IsNull) should add a constraint!
                 }
-                self.action(input, columns, gets);
+
+                // Special case for IsNull: check if the requirements in order for
+                // the inner expression to be null contradict the non-null
+                // requirements pushed down or derived from the rest of the
+                // predicates.
+                if predicates.iter().any(|p| {
+                    if let MirScalarExpr::CallUnary {
+                        expr: inner,
+                        func: UnaryFunc::IsNull,
+                    } = p
+                    {
+                        let mut required_null_cols = HashSet::new();
+                        inner.non_null_requirements(&mut required_null_cols);
+                        if required_null_cols.iter().any(|c| columns.contains(c)) {
```

Comment:
This seems incorrect. On this branch, we have
```
materialize=> explain select * from (select * from t1 where f1 + f2 is null) blah join t2 on blah.f1 = t2.f1;
 Optimized Plan 
----------------
 %0 =          +
 | Constant    +
```
The join condition results in a `t1.f1 is not null` restriction being pushed down. But `t1.f1 is not null and (t1.f1 + t1.f2 is null)` is not an impossible condition.

Reply:
Awesome catch! :facepalm:

Reply:
My fist instinct was to use `all` instead of `any`... But that doesn't work either as shown below

```
materialize=> explain select * from (select * from t1 where f1 + f2 is null) blah join t2 on blah.f1 = t2.f1;
                Optimized Plan                
----------------------------------------------
 %0 =                                        +
 | Get materialize.public.t1 (u788)          +
 | Filter !(isnull(#0)), isnull(#1)          +
 | ArrangeBy (#0)                            +
                                             +
 %1 =                                        +
 | Get materialize.public.t2 (u786)          +
 | Filter !(isnull(#0))                      +
                                             +
 %2 =                                        +
 | Join %0 %1 (= #0 #2)                      +
 | | implementation = Differential %1 %0.(#0)+
 | | demand = (#0, #1, #3)                   +
 | Project (#0, #1, #0, #3)                  +
 
(1 row)

materialize=> explain select * from (select * from t1 where f1 + coalesce(f1, f2) is null) blah join t2 on blah.f1 = t2.f1;
 Optimized Plan 
----------------
 %0 =          +
 | Constant    +
 
(1 row)

materialize=>
```

Reply:
I was trying to re-engineer the wheel here. We already have code for that in `canonicalize_predicates` I believe. I should just pushed that inferred info down as a predicate and let `canonicalize_predicates` do its job.

This has been my first attempt:

```
diff --git a/src/transform/src/nonnull_requirements.rs b/src/transform/src/nonnull_requirements.rs                                                                                            
index 9bc1fdc52..d72055c76 100644                                                                                                                                                             
--- a/src/transform/src/nonnull_requirements.rs                                                                                                                                               
+++ b/src/transform/src/nonnull_requirements.rs                                                                                                                                               
@@ -148,30 +148,7 @@ impl NonNullRequirements {                                                                                                                                               
                     // TODO: Not(IsNull) should add a constraint!                                                                                                                            
                 }                                                                                                                                                                            
                                                                                                                                                                                              
-                // Special case for IsNull: check if the requirements in order for                                                     
-                // the inner expression to be null contradict the non-null                                                             
-                // requirements pushed down or derived from the rest of the                                                            
-                // predicates.                                                                                                         
-                if predicates.iter().any(|p| {                                                                                         
-                    if let MirScalarExpr::CallUnary {                                                                                  
-                        expr: inner,                                                                                                   
-                        func: UnaryFunc::IsNull,                                                                                       
-                    } = p                                                                                                              
-                    {                                                                                                                  
-                        let mut required_null_cols = HashSet::new();                                                                   
-                        inner.non_null_requirements(&mut required_null_cols);                                                          
-                        if required_null_cols.iter().any(|c| columns.contains(c)) {                                                    
-                            return true;                                                                                               
-                        }                                                                                                              
-                    }                                                                                                                  
-                    false                                                                                                              
-                }) {                                                                                                                   
-                    // Impossible condition detected: a set of columns are required                                                    
-                    // to be null and non-null.                                                                                        
-                    relation.take_safely();                                                                                            
-                } else {                                                                                                               
-                    self.action(input, columns, gets);                                                                                 
-                }                                                                                                                      
+                self.action(input, columns, gets);                                                                                     
             }                                                                                                                          
             MirRelationExpr::Join {                                                                                                    
                 inputs,                                                                                                                
@@ -212,6 +189,15 @@ impl NonNullRequirements {                                                                                         
                 }                                                                                                                      
                                                                                                                                        
                 for (input, columns) in inputs.iter_mut().zip(new_columns) {                                                           
+                    if !columns.is_empty() {                                                                                           
+                        *input = input.take_dangerous().filter(columns.iter().map(|col| {                                              
+                            MirScalarExpr::Column(*col)                                                                                
+                                .call_unary(UnaryFunc::IsNull)                                                                         
+                                .call_unary(UnaryFunc::Not)                                                                            
+                        }));                                                                                                           
+                        // fuse nested filters, canonicalize predicates and detect impossible conditions                                                                                                                                                                       
+                        crate::fusion::filter::Filter.action(input);                                                                   
+                    }                                                                                                                  
                     self.action(input, columns, gets);                                                                                 
                 }                                                                                                                      
             }
```

However, that leads to a bunch of redundant IS NOT NULL predicates in a bunch of tests:

```
--- a/test/sqllogictest/transform/dataflow.slt                                                                                          
+++ b/test/sqllogictest/transform/dataflow.slt                                                                                          
@@ -26,7 +26,7 @@ EXPLAIN PLAN FOR SELECT * from foo3                                                                                   
 ----                                                                                                                                   
 %0 =                                                                                                                                   
 | Get materialize.public.foo (u1)                                                                                                      
-| Filter (#0 = 5), (#1 = 6)                                                                                                            
+| Filter !(isnull(#0)), !(isnull(#1)), (#0 = 5), (#1 = 6)                                                                              
 | Project (#1)                                                                                                                         
                                                                                                                                        
 EOF                                                                                                                                    
@@ -36,7 +36,7 @@ EXPLAIN PLAN FOR VIEW foo3                                                                                            
 ----                                                                                                                                   
 %0 =                                                                                                                                   
 | Get materialize.public.foo (u1)                                                                                                      
-| Filter (#0 = 5), (#1 = 6)                                                                                                            
+| Filter !(isnull(#0)), !(isnull(#1)), (#0 = 5), (#1 = 6)                                                                              
 | Project (#1)                                                                                                                         
                                                                                                                                        
 EOF                                                                                                                                    
@@ -49,7 +49,7 @@ EXPLAIN PLAN FOR VIEW foo3                                                                                            
 ----                                                                                                                                   
 %0 =                                                                                                                                   
 | Get materialize.public.foo2 (u3)                                                                                                     
-| Filter (#0 = 6)                                                                                                                      
+| Filter !(isnull(#0)), (#0 = 6)                                                                                                       
                                                                                                                                        
 EOF
...
```

I guess we could extend `canonicalize_predicates` to remove redudant `IS NOT NULL` predicates that are implied by the rest of predicates.

Reply:
yes we could.

Reply:
though we may want to hold off on extending canonicalize predicates until we can create arbitrary `MirScalarExpr`s that way it is easier to write unit tests.